api/lambda/authorizer.js

```diff
@@ -315,70 +318,128 @@ AuthPolicy.prototype = (function AuthPolicyClass() {
+  }
 
-  // you can send a 401 Unauthorized response to the client by failing like so:
-  // callback('Unauthorized');
+  //Get the kid from the token and retrieve corresponding PEM
+  var kid = decodedJwt.header.kid;
+  var pem = PEMS[kid];
+  if (!pem) {
+    console.log('Invalid Identity token');
+    callback("Unauthorized");
+    return;
+  }
 
-  // build apiOptions for the AuthPolicy
-  const apiOptions = {};
-  const tmp = event.methodArn.split(':');
-  const apiGatewayArnTmp = tmp[5].split('/');
-  const awsAccountId = tmp[4];
-  apiOptions.region = tmp[3];
-  apiOptions.restApiId = apiGatewayArnTmp[0];
-  apiOptions.stage = apiGatewayArnTmp[1];
+  //Verify the signature of the JWT token to ensure it's really coming from your User Pool
 
-  // this function must generate a policy that is associated with the recognized principal user identifier.
-  // depending on your use case, you might store policies in a DB, or generate them on the fly
+  jwt.verify(token, pem, {issuer: tokenIssuer}, function (err, payload) {
+    if (err) {
+      callback("Unauthorized");
+    } else {
+      //Valid token. Generate the API Gateway policy for the user
+      //Always generate the policy on value of 'sub' claim and not for
+      // 'username' because username is reassignable
+      //sub is UUID for a user which is never reassigned to another user.
 
-  // keep in mind, the policy is cached for 5 minutes by default (TTL is configurable in the authorizer)
-  // and will apply to subsequent calls to any method/resource in the RestApi
-  // made with the same token
+      const apiOptions = {};
+      const tmp = event.methodArn.split(':');
+      const apiGatewayArnTmp = tmp[5].split('/');
+      const principalId = payload.sub;
+      const awsAccountId = tmp[4];
 
-  // the example policy below denies access to all resources in the RestApi
-  const policy = new AuthPolicy(principalId, awsAccountId, apiOptions);
+      apiOptions.region = tmp[3];
+      apiOptions.restApiId = apiGatewayArnTmp[0];
+      apiOptions.stage = apiGatewayArnTmp[1];
 
-  // Add generic allow for any valid user, with the following DENY statements taking precedence when included
-  policy.allowAllMethods();
+      let admin = null;
+      let policy = new AuthPolicy(principalId, awsAccountId, apiOptions);
+      policy.allowAllMethods();
 
 
-  if (!admin) {
-    policy.denyMethod(AuthPolicy.HttpVerb.DELETE, '/locations');
-    policy.denyMethod(AuthPolicy.HttpVerb.DELETE, '/locations/*');
-    policy.denyMethod(AuthPolicy.HttpVerb.POST, '/locations');
-    policy.denyMethod(AuthPolicy.HttpVerb.POST, '/locations/*');
+      //Check the Cognito group entry for Admin.
+      //Assuming here that the Admin group has always higher
+      //precedence
+
+      if (payload['cognito:groups'] &&
+        payload['cognito:groups'][0] === 'adminGroup') {
+        admin = true;
+      }
+
+      if (!admin) {
+        policy.denyMethod(AuthPolicy.HttpVerb.DELETE, '/locations');
```

Comment:
Even when I logged in with "admin1," and it showed that I was a member of the adminGroup in the logs, I found that the effective policy generated by the custom authorizer denied access to admin-only API operations.